File: backend/src/main/kotlin/com/example/backend/controllers/CaisseController.kt
package com.example.backend.controllers

import com.example.backend.dtos.CaisseClotureRequestDto
import com.example.backend.dtos.CaisseOuvertureRequestDto
import com.example.backend.repositories.CaisseRepository
import com.example.backend.repositories.EmployeRepository
import com.example.backend.repositories.UserRepository
import com.example.backend.services.CaisseException
import com.example.backend.services.CaisseService
import com.example.backend.utility.UserUtils
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/caisses")
class CaisseController(
  private val userUtils: UserUtils,
  private val caisseService: CaisseService,
  private val employeRepository: EmployeRepository,
  private val caisseRepository: CaisseRepository,
  private val userRepository: UserRepository
) {

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/active/details")
  fun getActiveCaisseDetails(): ResponseEntity<Map<String, Any?>> {
    return try {
      val activeCaisse = caisseService.getCaisseActive()
      if (activeCaisse != null) {
        val employeName = activeCaisse.user?.user?.nom ?: "Inconnu"
        val caisseDetails = mapOf(
          "id" to activeCaisse.id,
          "etat" to activeCaisse.etat,
          "nomEmploye" to employeName,
          "dateOuvert" to activeCaisse.dateOuvert,
          "dateFerme" to activeCaisse.dateFerme
        )
        ResponseEntity.ok(caisseDetails)
      } else {
        ResponseEntity.status(HttpStatus.NOT_FOUND).body(mapOf("error" to "Aucune caisse active trouvée"))
      }
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Une erreur interne est survenue: ${e.message}"))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/active/details/pageable")
  fun getActiveCaissesDetailsPageable(
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sortBy: String
  ): ResponseEntity<Page<Map<String, Any?>>> {
    return try {
      val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, sortBy))
      val caisseDetails = caisseService.getAllCaisse(pageable)
      ResponseEntity.ok(caisseDetails)
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Page.empty())
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/ouverte")
  fun isCaisseOuverte(): ResponseEntity<Map<String, Any?>> {
    val caisseActive = caisseService.getCaisseActive()
    val caisseEnCours = caisseService.getCaisseEnCours()
    val lastCaisse = caisseService.getLastCaisse()
    val employeCurrentId = userUtils.getCurrentEmployeId()
    var employeCurrent = employeRepository.findById(employeCurrentId!!.toInt()).get()

    val caisseEnCoursCurrentUser =
      caisseRepository.findByUserAndEtatAndSupprimer(employeCurrent, "En cours", 0).firstOrNull()


    val response: Map<String, Any?> =
      if (lastCaisse?.etat == "Clot") {
        mapOf(
          "status" to "close",
          "caisseDetails" to null
        )
      } else
        if (caisseActive != null && caisseActive.user?.id == employeCurrent?.id?.toInt()) {
          mapOf(
            "status" to "active",
            "caisseDetails" to mapOf(
              "id" to caisseActive.id,
              "etat" to caisseActive.etat,
              "session" to caisseActive.session,
              "nomEmploye" to (caisseActive.user?.user?.nom ?: "Inconnu"),
              "dateOuvert" to caisseActive.dateOuvert,
              "dateFerme" to caisseActive.dateFerme
            )
          )
        } else if (employeCurrent != null && caisseEnCoursCurrentUser != null && caisseEnCoursCurrentUser?.user?.id == employeCurrent.id?.toInt()) {
          mapOf(
            "status" to "pending",
            "caisseDetails" to mapOf(
              "id" to caisseEnCoursCurrentUser?.id,
              "etat" to caisseEnCoursCurrentUser?.etat,
              "session" to caisseEnCoursCurrentUser?.session,
              "nomEmploye" to (caisseEnCoursCurrentUser?.user?.user?.nom ?: "Inconnu"),
              "dateOuvert" to caisseEnCoursCurrentUser?.dateOuvert,
              "dateFerme" to caisseEnCoursCurrentUser?.dateFerme
            )
          )
        } else if (caisseActive != null && caisseEnCoursCurrentUser != null) {
          mapOf(
            "status" to "already",
            "caisseDetails" to mapOf(
              "id" to caisseActive.id,
              "etat" to caisseActive.etat,
              "session" to caisseActive.session,
              "nomEmploye" to (caisseActive.user?.user?.nom ?: "Inconnu"),
              "dateOuvert" to caisseActive.dateOuvert,
              "dateFerme" to caisseActive.dateFerme
            )
          )
        } else if (caisseActive == null && caisseEnCours != null) {
          mapOf(
            "status" to "open",
            "caisseDetails" to mapOf(
              "id" to caisseActive?.id,
              "etat" to caisseActive?.etat,
              "session" to caisseActive?.session,
              "nomEmploye" to (caisseActive?.user?.user?.nom ?: "Inconnu"),
              "dateOuvert" to caisseActive?.dateOuvert,
              "dateFerme" to caisseActive?.dateFerme
            )
          )
        } else if (caisseActive != null && caisseEnCours == null) {
          mapOf(
            "status" to "already",
            "caisseDetails" to mapOf(
              "id" to caisseActive?.id,
              "etat" to caisseActive?.etat,
              "session" to caisseActive?.session,
              "nomEmploye" to (caisseActive?.user?.user?.nom ?: "Inconnu"),
              "dateOuvert" to caisseActive?.dateOuvert,
              "dateFerme" to caisseActive?.dateFerme
            )
          )
        } else if (caisseActive == null) {
          mapOf(
            "status" to "close",
            "caisseDetails" to null
          )
        } else {
          mapOf(
            "status" to "close",
            "caisseDetails" to null
          )
        }


    return ResponseEntity.ok(response)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PutMapping("/active/attente-cloture")
  fun setCaisseToPendingClosure(): ResponseEntity<Map<String, Any?>> {
    return try {
      val updatedCaisse = caisseService.setCaisseToPendingClosure()
      ResponseEntity.ok(mapOf("message" to "La caisse a été mise en attente de clôture.", "caisse" to updatedCaisse))
    } catch (e: CaisseException) {
      ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Une erreur interne est survenue: ${e.message}"))
    }
  }


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/cloturer")
  fun cloturerCaisse(
    @RequestBody caisseClotureRequest: CaisseClotureRequestDto
  ): ResponseEntity<Any> {
    return try {
      val caisseDto =
        caisseService.cloturerCaisse(caisseClotureRequest.fondCaisseFerme, caisseClotureRequest.fermetureCaisse)
      ResponseEntity.ok(caisseDto)
    } catch (e: CaisseException) {
      ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Une erreur interne est survenue: ${e.message}"))
    }
  }


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/en_cours")
  fun mettreCaisseEnAttente(): ResponseEntity<Any> {
    return try {
      val caisseDto = caisseService.mettreCaisseEnAttente()
      ResponseEntity.ok(caisseDto)
    } catch (e: CaisseException) {
      ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Une erreur interne est survenue: ${e.message}"))
    }
  }


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/ouvrir")
  fun ouvrirNouvelleCaisse(
    @RequestBody caisseOuvertureRequest: CaisseOuvertureRequestDto
  ): ResponseEntity<Any> {
    return try {
      val caisseDto = caisseService.ouvrirNouvelleCaisse(caisseOuvertureRequest)
      ResponseEntity.ok(caisseDto)
    } catch (e: CaisseException) {
      ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Une erreur interne est survenue: ${e.message}"))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/cloture/details")
  fun getCaisseClosureDetails(): ResponseEntity<Any> {
    return try {
      val caisseDetails = caisseService.getCaisseAttenteCloture()
      if (caisseDetails != null) {
        ResponseEntity.ok(caisseDetails)
      } else {
        ResponseEntity.status(HttpStatus.NOT_FOUND)
          .body(mapOf("error" to "Aucune caisse en attente de clôture trouvée"))
      }
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
        .body(mapOf("error" to "Une erreur interne est survenue: ${e.message}"))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{caisseId}/rapport")
  fun getCaisseReport(@PathVariable caisseId: Long): ResponseEntity<Map<String, Any?>> {
    return try {
      val reportData = caisseService.generateCaisseReport(caisseId)
      ResponseEntity.ok(reportData)
    } catch (e: Exception) {
      ResponseEntity.status(500).body(mapOf("error" to "An internal error occurred: ${e.message}"))
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/all/pageable")
  fun getAllCaisses(
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sortBy: String
  ): ResponseEntity<Page<Map<String, Any?>>> {
    return try {
      val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, sortBy))
      val caisses = caisseService.getAllCaisses(pageable)
      ResponseEntity.ok(caisses)
    } catch (e: Exception) {
      ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Page.empty())
    }
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/all/filter")
  fun getFilteredCaisses(
    @RequestParam(required = false) caisseId: Long?,
    @RequestParam(required = false) startDate: String?,
    @RequestParam(required = false) endDate: String?,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int
  ): ResponseEntity<Page<Map<String, Any?>>> {
    val pageable = PageRequest.of(page, size)
    val caisses = caisseService.getFilteredCaisses(caisseId, startDate, endDate, pageable)
    return ResponseEntity.ok(caisses)
  }

}

File: backend_py/pharmaPython/app/api/v1/caisse_controller.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional, Dict, Any, List

from app.api.deps import get_db
from app.models.employe import Employe
from app.schemas.caisse_dto import CaisseClotureRequestDto, CaisseOuvertureRequestDto
from app.services.caisse_service import CaisseService
from app.utility.jwt_authentication import jwt_authentication
from app.utility.user_utils import UserUtils

router = APIRouter(
  prefix="/caisses",
  tags=["Caisse"],
  dependencies=[Depends(jwt_authentication)],
)


# ----------------------------
# Détails de la caisse active
# ----------------------------
@router.get("/active/details")
def get_active_caisse_details(db: Session = Depends(get_db)):
  service = CaisseService(db)
  active_caisse = service.get_caisse_active()
  if active_caisse:
    employe_name = active_caisse.user.user.nom if active_caisse.user else "Inconnu"
    return {
      "id": active_caisse.id,
      "etat": active_caisse.etat,
      "nomEmploye": employe_name,
      "dateOuvert": active_caisse.date_ouvert,
      "dateFerme": active_caisse.date_ferme
    }
  raise HTTPException(status_code=404, detail="Aucune caisse active trouvée")


# ----------------------------
# Vérifier si une caisse est ouverte
# ----------------------------
@router.get("/ouverte")
def is_caisse_ouverte(db: Session = Depends(get_db), employe: Employe = Depends(UserUtils.get_current_employe)):
  service = CaisseService(db)
  caisse_active = service.get_caisse_active()
  caisse_en_cours = service.get_caisse_en_cours()
  last_caisse = service.get_last_caisse()
  employe_current_id = employe.id

  # ⚠️ Ici la logique détaillée Kotlin est très longue → à mapper selon besoin
  if last_caisse and last_caisse.etat == "Clot":
    return {"status": "close", "caisseDetails": None}
  elif caisse_active and caisse_active.user_id == employe_current_id:
    return {"status": "active", "caisseDetails": service.map_to_dto(caisse_active)}
  elif caisse_active:
    return {"status": "already", "caisseDetails": service.map_to_dto(caisse_active)}
  elif caisse_en_cours:
    return {"status": "open", "caisseDetails": service.map_to_dto(caisse_en_cours)}
  return {"status": "close", "caisseDetails": None}


# ----------------------------
# Mettre la caisse en attente de clôture
# ----------------------------
@router.put("/active/attente-cloture")
def set_caisse_to_pending_closure(db: Session = Depends(get_db),
                                  employe: Employe = Depends(UserUtils.get_current_employe)):
  service = CaisseService(db)
  try:
    updated_caisse = service.set_caisse_to_pending_closure(employe)
    return {"message": "La caisse a été mise en attente de clôture.", "caisse": updated_caisse}
  except Exception as e:
    raise HTTPException(status_code=409, detail=str(e))


# ----------------------------
# Clôturer une caisse
# ----------------------------
@router.post("/cloturer")
def cloturer_caisse(request: CaisseClotureRequestDto, db: Session = Depends(get_db),
                    employe: Employe = Depends(UserUtils.get_current_employe)):
  service = CaisseService(db)
  try:
    caisse_dto = service.cloturer_caisse(request.fond_caisse_ferme, request.fermeture_caisse, employe)
    return caisse_dto
  except Exception as e:
    raise HTTPException(status_code=409, detail=str(e))


# ----------------------------
# Mettre une caisse en attente (en cours)
# ----------------------------
@router.get("/en_cours")
def mettre_caisse_en_attente(db: Session = Depends(get_db), employe: Employe = Depends(UserUtils.get_current_employe)):
  service = CaisseService(db)
  try:
    caisse_dto = service.mettre_caisse_en_attente(employe)
    return caisse_dto
  except Exception as e:
    raise HTTPException(status_code=409, detail=str(e))


# ----------------------------
# Ouvrir une nouvelle caisse
# ----------------------------
@router.post("/ouvrir")
def ouvrir_nouvelle_caisse(request: CaisseOuvertureRequestDto, db: Session = Depends(get_db),
                           employe: Employe = Depends(UserUtils.get_current_employe)):
  service = CaisseService(db)
  # try:
  caisse_dto = service.ouvrir_nouvelle_caisse(request, employe)
  return caisse_dto
  # except Exception as e:
  #   raise HTTPException(status_code=409, detail=str(e))


# ----------------------------
# Récupérer détails d'une caisse en attente de clôture
# ----------------------------
@router.get("/cloture/details")
def get_caisse_closure_details(db: Session = Depends(get_db)):
  service = CaisseService(db)
  caisse_details = service.get_caisse_attente_cloture()
  if caisse_details:
    return caisse_details
  raise HTTPException(status_code=404, detail="Aucune caisse en attente de clôture trouvée")


# ----------------------------
# Rapport d'une caisse
# ----------------------------
@router.get("/{caisse_id}/rapport")
def get_caisse_report(caisse_id: int, db: Session = Depends(get_db)):
  service = CaisseService(db)
  return service.generate_caisse_report(caisse_id)


# ----------------------------
# Toutes les caisses (pagination)
# ----------------------------
@router.get("/all/pageable")
def get_all_caisses(page: int = Query(0), size: int = Query(10), db: Session = Depends(get_db)):
  service = CaisseService(db)
  return service.get_all_caisses(page=page, size=size)


# ----------------------------
# Filtrer les caisses (pagination)
# ----------------------------
@router.get("/all/filter")
def get_filtered_caisses(
  caisse_id: Optional[int] = None,
  start_date: Optional[str] = None,
  end_date: Optional[str] = None,
  page: int = Query(0),
  size: int = Query(10),
  db: Session = Depends(get_db)
):
  service = CaisseService(db)
  return service.get_filtered_caisses(caisse_id, start_date, end_date, page=page, size=size)
